inmem/kvlock.go

```diff
@@ -0,0 +1,150 @@
+package inmem
+
+import "context"
+
+//  This is an RWMutex using channels that allows for context timeouts.
+//  It is adapted from Rob Pike's blog post:
+//  https://blogtitle.github.io/go-advanced-concurrency-patterns-part-3-channels/
+// where he made an RWMutex using channels.  I just added the context handlng.
+// h/t to Bryan Mills for pointing out this blog post to me.
+type RWMutex struct {
+	write   chan struct{}
+	readers chan int
+}
+
+func NewLock() RWMutex {
+	return RWMutex{
+		// This is used as a normal Mutex.
+		write: make(chan struct{}, 1),
+		// This is used to protect the readers count.
+		// By receiving the value it is guaranteed that no
+		// other goroutine is changing it at the same time.
+		readers: make(chan int, 1),
+	}
+}
+
+func (l RWMutex) Lock() { l.write <- struct{}{} }
+
+func (l RWMutex) TryLock(ctx context.Context) bool {
+	if ctx == nil {
+		select {
+		case l.write <- struct{}{}:
+			return true
+		default:
+			return false
+
+		}
+	}
+	done := ctx.Done()
+	select {
+	case l.write <- struct{}{}:
+		select {
+		case <-done:
+			<-l.write
```

Comment:
This is not clear to me 🤔 

Why would you grab the lock if the context is done?
Should you release instead

```
l.write <- struct{}{}
```
?

Reply:
That isn't a release.  `<-l.write` is a release
`l.write <- struct{}{}` grabs the write lock.

Reply:
Sure thing, I got confused 😅